pyro/contrib/funsor/named_messenger.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import OrderedDict, namedtuple
+from contextlib import ExitStack  # python 3
+
+from pyro.poutine.reentrant_messenger import ReentrantMessenger
+
+
+# name_to_dim : dict, dim_to_name : dict
+StackFrame = namedtuple('StackFrame', ['name_to_dim', 'dim_to_name'])
+
+
+class NamedMessenger(ReentrantMessenger):
+    """
+    Handler for converting to/from funsors consistent with Pyro's positional batch dimensions.
+
+    :param int history: The number of previous contexts visible from the
+        current context. Defaults to 1. If zero, this is similar to
+        :class:`pyro.plate`.
+    :param bool keep: If true, frames are replayable. This is important
+        when branching: if ``keep=True``, neighboring branches at the same
+        level can depend on each other; if ``keep=False``, neighboring branches
+        are independent (conditioned on their shared ancestors).
+    """
+    def __init__(self, history=1, keep=False):
+        assert history >= 0
+        self.history = history
+        self.keep = keep
+        self._iterable = None
+        self._pos = -1
+        self._stack = []
+        super().__init__()
+
+    def generator(self, iterable):
+        self._iterable = iterable
+        return self
+
+    def __iter__(self):
+        with ExitStack() as stack:
+            for value in self._iterable:
+                stack.enter_context(self)
+                yield value
+
+    def __enter__(self):
+        self._pos += 1
+        if len(self._stack) <= self._pos:
+            self._stack.append(
+                StackFrame(name_to_dim=OrderedDict(), dim_to_name=OrderedDict()))
+        return super().__enter__()
+
+    def __exit__(self, *args, **kwargs):
+        if not self.keep:
+            self._stack.pop()
+        self._pos -= 1
+        return super().__exit__(*args, **kwargs)
+
+    @property
+    def current_frame(self):
+        return self._stack[self._pos]
+
+    def _pyro_to_data(self, msg):
+        if len(msg["args"]) < 2:
+            msg["args"] = msg["args"] + (OrderedDict(),)
+        funsor_value, name_to_dim = msg["args"]
+
+        # every markov context is responsible for knowing about non-fresh local vars,
+        # and fresh vars are handled by passing information up & down the stack
+
+        # handling non-fresh vars: just look at the deepest context they appear in
+        name_to_dim.update({name: self.current_frame.name_to_dim[name]
+                            for name in funsor_value.inputs
+                            if name in self.current_frame.name_to_dim
+                            and name not in name_to_dim})
```

Comment:
`NamedMessenger._pyro_to_data` populates a `name_to_dim` mapping for a new `pyro.contrib.funsor.to_data` call according to all previous `pyro.contrib.funsor.to_data` calls it or its parents have seen so far.